### PR TITLE
Improves solution with waypoint optimization using gradient descent

### LINKED ISSUE #41: Update mothership waypoint calc
Ensure suitable waypoints that reflect effective use of resources by considering:
- radial distance from centroid, or
- other.

Consider cases where clusters are close together, i.e., already within deployment radius.

Image

### PR DESCRIPTION
**Pull request by ryu-AIMS**
Implements a waypoint optimization heuristic within the initial solution generation.

- Introduces a gradient descent method to refine waypoint positions, aiming to reduce overall path cost - a trade-off between the mothership and tenders distances.
- Ensures tender sorties are updated to accommodate the adjusted waypoints, maintaining feasibility.

Closes #41

### CODE REVIEW

File: src/routing/routing_heuristics.jl [hunk 1]
```diff
@@ -227,6 +227,404 @@ function get_waypoints(
...
+    problem::Problem;
+    learning_rate::Float64=1e-8,
+    tolerance::Float64=1e-6,
+    max_desc_iters::Int=50,
+    max_search_iters::Int=10
+)::MSTSolution
+    exclusions_mothership::DataFrame = problem.mothership.exclusion
+    exclusions_tender::DataFrame = problem.tenders.exclusion
+    n_tenders::Int8 = problem.tenders.number
+    t_cap::Int16 = problem.tenders.capacity
```
[the remark this answers is not in the document]
Comment (ryu-AIMS) [reply src/routing/routing_heuristics.jl:266]: Just wanted to assign (single-name) aliases to reference fields in problem struct, avoiding re-writing every time.

File: src/routing/routing_heuristics.jl [hunk 2]
```diff
@@ -227,6 +227,404 @@ function get_waypoints(
...
+    vessel_weightings::NTuple{2,AbstractFloat} = (
+        problem.mothership.weighting, problem.tenders.weighting
+    )
+    wpts = copy(soln.mothership_routes[end].route.nodes)
+    n = length(wpts)
```
[the remark this answers is not in the document]
Comment (ryu-AIMS) [reply src/routing/routing_heuristics.jl]: [0a35812](https://github.com/ryu-AIMS/HierarchicalRouting.jl/pull/69/commits/0a3581282cce59b8388a2f47cce6773741343261)

File: src/routing/routing_heuristics.jl [hunk 3]
```diff
@@ -227,6 +227,404 @@ function get_waypoints(
...
+- `learning_rate`: Step size for the descent.
+- `tolerance`: Tolerance for stopping the descent.
+- `max_desc_iters`: Maximum number of descent iterations.
+- `δ`: Perturbation step size for gradient calculation.
+- `vessel_weightings`: Weightings for mothership and tender costs.
+
+# Returns
+The updated waypoint after the descent step.
+"""
+function waypoint_descent_step(
```
[the remark this answers is not in the document]
Comment (ryu-AIMS) [reply src/routing/routing_heuristics.jl:359]: Not sure what you mean. 
Test performance/solution quality to see if it performs as intended?

File: src/routing/routing_heuristics.jl [hunk 4]
```diff
@@ -227,6 +227,404 @@ function get_waypoints(
...
+- `vessel_weightings`: Weightings for mothership and tender costs.
+
+# Returns
+- The (critical path) cost of the proposed solution.
+- The updated MSTSolution with the proposed waypoint.
+"""
+function evaluate_perturbation(
+    point_proposed::Point{2,Float64},
+    idx::Int64,
+    waypoints_ex::Vector{Point{2,Float64}},
```
[the remark this answers is not in the document]
Comment (ConnectedSystems) [reply src/routing/routing_heuristics.jl:427]: My two cents: conventions can be established but must balance with readability of the code.

I don't mind if it's an established convention in the field, e.g., $Q$ for flux, $\hat{y}$ (`y_hat`) for emulated or predicted value, and I agree that you can establish your own convention within a codebase.

That said, `existing_waypoints` is more readable (`waypoints_existing` slightly less so). 
Note you use `waypoints_proposed` anyway so it "feels" unbalanced in a way (why use shorthand for one but not the other?).

You also ignore an existing convention in this codebase (`solution` -> `soln`).
So `solution_ex` could be `existing_soln` which, again to me, is more readable.

File: src/routing/routing_heuristics.jl [hunk 5]
```diff
@@ -227,6 +227,414 @@ function get_waypoints(
...
+    )
+
+    return solution_opt
+end
+
+"""
+    waypoint_descent_step(
+        idx::Int,
+        wpts::Vector{Point{2,Float64}},
+        soln::MSTSolution,
```
Comment (ConnectedSystems) [inline src/routing/routing_heuristics.jl:331]: Did you have an established convention of always having `soln` as the first argument, or am I misremembering?

I would find it annoying having to remember where in the argument position `soln` is meant to be across all the methods you have if the expected position keeps changing.

File: src/routing/routing_heuristics.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryu-AIMS) [reply src/routing/routing_heuristics.jl]: [2ac5635](https://github.com/ryu-AIMS/HierarchicalRouting.jl/pull/69/commits/2ac56355c80cad89eaa9bf54b363d3a7cef80cc7)

File: src/routing/routing_heuristics.jl (hunk 5 above)
Comment (ryu-AIMS) [reply src/routing/routing_heuristics.jl]: [e5c4e02](https://github.com/ryu-AIMS/HierarchicalRouting.jl/pull/69/commits/e5c4e021a3c0f485772cf15dd6ace7318ca8ea4f)

File: src/routing/routing_heuristics.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (Zapiano) [reply src/routing/routing_heuristics.jl:361]: I mean test the result. If you apply gradient descent to a known function, such as `y = x^2 + 5` to find it's minima, you know what the result should be, so you can test to see if your implementation of gradient descent is working as expected (at least in known cases). That doesn't guarantee it is always working, but at least adds a level of trust.
